akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala

```diff
@@ -120,14 +146,17 @@ import scala.util.control.NonFatal
   }
 
   override final def onSubscribe(s: Subscription): Unit = {
-
-    @tailrec def rec(obj: AnyRef): Unit =
+    @tailrec def rec(obj: AnyRef): Unit = {
       get() match {
-        case null ⇒ if (!compareAndSet(null, obj)) rec(obj)
+        case null ⇒
+          if (VirtualProcessor.Debug) println(s"VirtualPublisher#$hashCode(null).onSubscribe.rec($s) -> ${obj.getClass}")
+          if (!compareAndSet(null, obj)) rec(obj)
         case subscriber: Subscriber[_] ⇒
+          if (VirtualProcessor.Debug) println(s"VirtualPublisher#$hashCode($subscriber).onSubscribe.rec($s) moving to both")
           obj match {
             case subscription: Subscription ⇒
-              if (compareAndSet(subscriber, Both.create(subscriber))) establishSubscription(subscriber, subscription)
+              val establishing = Establishing.create(subscriber)
+              if (compareAndSet(subscriber, Both.create(subscriber))) establishSubscription(establishing, subscription)
```

Comment:
What would happen before is that the state was set to `Both` here, but before `establishSubscription` would get to call `subscriber.onSubscribe` `onComplete` would execute and complete the sub, which would then get an `onSubscribe` call after being completed (that's the spec violation bug).